Sources/Workspace/Destination.swift

```diff
@@ -84,7 +98,7 @@ public struct Destination {
         environment: [String:String] = Process.env
     ) throws -> Destination {
         // Select the correct binDir.
-        let binDir = binDir ?? Destination.hostBinDir(
+        let binDir = try binDir ?? Destination.hostBinDir(
```

Comment:
Sorry it took so long but I finally found some time to think about this PR. Here are my thoughts:

I don't think it's a good idea to bake in some heuristics about what makes the toolchain "complete". Many tools in the toolchain can be considered implementation detail and they may get removed once they serve no purpose. For example, swift-build-tool is no longer required in the toolchain. What if we make `binDir` an enum so clients can control _how_ the binDir is found. Something like this:

```swift
enum BinDirSearchMode {
    /// SwiftPM is being used as a package. Look the toolchain according to the host platform. 
    case hostLookup

    /// SwiftPM is inside the toolchain and should lookup bin dir relative to its path.
    case toolchain

    /// A custom bin directory is provided.
    case custom(AbsolutePath)
}
```

Ofc, the names could be better but you get the general idea. Clients who are using SwiftPM as a package would be able to pass in hostLookup or a custom bin directory if they wish. The hostLookup case would call UserToolchain's getHostToolchain method to locate the toolchain. What do you think?

Reply:
> I don’t think it's a good idea to bake in some heuristics about what makes the toolchain “complete”. Many tools in the toolchain can be considered implementation detail and they may get removed once they serve no purpose.

That is precisely the point of this proposal. Right now any client needs to implement some heuristic itself from the outside. As you say, they are implementation details which may be removed, and that is why I as a client do not what to have to depend on them from the outside anymore. SwiftPM is far better poised than outside clients to know that the implementation details are changing and to adjust the heuristic accordingly.

(In the time since I first made this pull request, I figured out that validation can be sidestepped by reversing the search order and going from most third‐party to most integral. Then the first one to be found is likely to be complete and not a shim. But it feels wrong and it ends up introducing the maximum—as opposed to the minimum—amount of interference between the SwiftPM library and its resources in the Swift toolchain. That in turn introduces fragility.)

***

While I have no real objection to it, I do not see what would be gained by expressing it as an enumeration. It would only duplicate API surface.

- The behaviour described by `custom` is already available now by providing the path to [the parent method](https://github.com/apple/swift-package-manager/blob/756eeaa41de0336639428275b0881fc73bfcf037/Sources/Workspace/Destination.swift#L82-L87).
- The behaviour of `toolchain` is what it does now when you specify no path.
- `hostLookup` is the only new one. The name doesn’t sound like it cares how the toolchain is found, only that whatever it does find works. Why wouldn’t it just be the fallback after the other two options have failed?

On the other hand maybe you want an internal way to assume the toolchain and prevent lookup for security reasons when you build as part of the Swift project? That would force us back in the direction of using `#if`.

Reply:
Right, `custom` and `toolchain` preserves the existing behavior and `hostLookup` allows automatically lookup of the toolchain. 

I might be missing something about your use case but from what I understand, you want libSwiftPM to automatically determine the toolchain when it's being used as a package dependency. With this approach, those would be able to create a toolchain with a single line if they don't need any customization. Something like:

```swift
let toolchain = UserToolchain.host
```

Is that not what you want?

Reply:
> Is that not what you want?

I think you are basically describing what I want, though the toolchain is not the ultimate goal. This is what I want to just work:

```swift
let graph = try PackageGraph(path: "/Users/John Doe/Repositories/SomePackage/")
```

...but also for `Manifest` and `Package` when the whole graph isn’t needed.

Reply:
...or if necessary, we drop this pull request and aim only slightly lower at this instead:

```swift
let graph = try PackageGraph(
    path: "/Users/John Doe/Repositories/SomePackage/",
    swiftToolchain: "/usr/bin/swift"
)
```

Reply:
We can make something like that work. Let me try to come up with a plan.